server/broker/store.go

```diff
@@ -0,0 +1,167 @@
+//limitations under the License.
+package broker
+
+import (
+	"sync"
+
+	"github.com/ServiceComb/service-center/pkg/async"
+	"github.com/ServiceComb/service-center/pkg/util"
+	sstore "github.com/ServiceComb/service-center/server/core/backend/store"
+	pb "github.com/ServiceComb/service-center/server/core/proto"
+)
+
+const (
+	PARTICIPANT sstore.StoreType = iota
+	VERSION
+	PACT
+	PACT_VERSION
+	PACT_TAG
+	VERIFICATION
+	PACT_LATEST
+	typeEnd
+)
+
+var TypeNames = []string{
+	PARTICIPANT:  "PARTICIPANT",
+	VERSION:      "VERSION",
+	PACT:         "PACT",
+	PACT_VERSION: "PACT_VERSION",
+	PACT_TAG:     "PACT_TAG",
+	VERIFICATION: "VERIFICATION",
+	PACT_LATEST:  "PACT_LATEST",
+}
+
+var TypeRoots = map[sstore.StoreType]string{
+	PARTICIPANT:  GetBrokerParticipantKey(""),
+	VERSION:      GetBrokerVersionKey(""),
+	PACT:         GetBrokerPactKey(""),
+	PACT_VERSION: GetBrokerPactVersionKey(""),
+	PACT_TAG:     GetBrokerTagKey(""),
+	VERIFICATION: GetBrokerVerificationKey(""),
+	PACT_LATEST:  GetBrokerLatestKey(""),
+}
+
+var store = &BKvStore{}
+
+func Store() *BKvStore {
+	return store
+}
+
+func (s *BKvStore) StoreSize(t sstore.StoreType) int {
+	return 100
+}
+
+func (s *BKvStore) dispatchEvent(t sstore.StoreType, evt *sstore.KvEvent) {
+	s.bindexers[t].OnCacheEvent(evt)
+	select {
+	case <-s.Ready():
+	default:
+		if evt.Action == pb.EVT_CREATE {
+			evt.Action = pb.EVT_INIT
+		}
+	}
+	sstore.EventProxy(t).OnEvent(evt)
+}
+
+func (s *BKvStore) newStore(t sstore.StoreType, opts ...sstore.KvCacherCfgOption) {
+	opts = append(opts,
+		sstore.WithKey(TypeRoots[t]),
+		sstore.WithInitSize(s.StoreSize(t)),
+		sstore.WithEventFunc(func(evt *sstore.KvEvent) { s.dispatchEvent(t, evt) }),
+	)
+	s.newIndexer(t, sstore.NewKvCacher(opts...))
+}
+
+func (s *BKvStore) store() {
+	for t := sstore.StoreType(0); t != typeEnd; t++ {
+		s.newStore(t)
+	}
+	for _, i := range s.bindexers {
+		<-i.Ready()
+	}
+	util.SafeCloseChan(s.bready)
+
+	util.Logger().Debugf("all indexers are ready")
+}
+
+func init() {
+	store.Initialize()
+
+	//sstore.AddEventHandleFunc(LEASE, store.onLeaseEvent)
+}
+
+type BKvStore struct {
+	*sstore.KvStore
+	bindexers     map[sstore.StoreType]*sstore.Indexer
+	basyncTaskSvc *async.AsyncTaskService
+	block         sync.RWMutex
+	bready        chan struct{}
+	bisClose      bool
+}
+
+func (s *BKvStore) Initialize() {
+	s.KvStore = sstore.Store()
+	s.KvStore.Initialize()
+	s.bindexers = make(map[sstore.StoreType]*sstore.Indexer)
+	s.basyncTaskSvc = async.NewAsyncTaskService()
+	s.bready = make(chan struct{})
+
+	for i := sstore.StoreType(0); i != typeEnd; i++ {
+		store.newNullStore(i)
+	}
+}
+
+func (s *BKvStore) newNullStore(t sstore.StoreType) {
+	s.newIndexer(t, sstore.NullCacher)
+}
+
+func (s *BKvStore) newIndexer(t sstore.StoreType, cacher sstore.Cacher) {
+	indexer := sstore.NewCacheIndexer(t, cacher)
+	s.bindexers[t] = indexer
+	indexer.Run()
+}
+
+func (s *BKvStore) Run() {
```

Comment:
This func does not been call at all, so the cache mechanism is not work here.

Reply:
fixed.